### PR TITLE
Refactor search

### PR DESCRIPTION
**Pull request by smashercosmo**
**Summary**
Search refactoring:
1. Convert to TS
2. Proper search action type
3. Replace immutable with immer
4. General cleanup

I made some debatable changes. Details are in comments.

**Test plan**
I tested both entries search and relations search manually

### CODE REVIEW

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:44]: Couldn't figure out why do we need to send the same data. If it's really needed, I can bring it back.

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:56]: Same here. searchTerm is already in state. we don't need to send it again.

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:72]: collection and searchFields are not used anywhere

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:113]: do we really need 'isFetching' check here?

File: packages/netlify-cms-core/src/reducers/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/reducers/search.ts]: Those kind of checks don't belong to reducer

File: packages/netlify-cms-core/src/reducers/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/reducers/search.ts]: fetchID is not used anywhere

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:123]: No reason to send all of this data in case of failure

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/search.ts:140]: can 'response.pagination' be different from passed 'page' param?

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (smashercosmo) [inline packages/netlify-cms-widget-relation/src/RelationControl.js:120]: I would say that this is not idiomatic usage of redux action. One shouldn't rely on the return value of the action and get all the data only from props.

File: packages/netlify-cms-core/src/actions/search.ts
Comment (erezrokah) [reply packages/netlify-cms-core/src/actions/search.ts:140]: When not using an integration (which is the default), `response.pagination` is missing from `response`

File: packages/netlify-cms-core/src/actions/search.ts
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/search.ts:17]: Nice!

File: packages/netlify-cms-core/src/actions/search.ts
Comment (erezrokah) [reply packages/netlify-cms-core/src/actions/search.ts:113]: I think it's useful to refresh the results (you have a succesful search and then search for the same thing).

File: packages/netlify-cms-core/src/reducers/search.ts
Comment (erezrokah) [inline packages/netlify-cms-core/src/reducers/search.ts:74]: I know this wasn't the case before, but should we scope `isFetching` and `term` to `namespace` too? (similar to `queryHits`)

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (erezrokah) [reply packages/netlify-cms-widget-relation/src/RelationControl.js:120]: Completely agree 💯

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (erezrokah) [inline packages/netlify-cms-widget-relation/src/RelationControl.js:122]: This has the potential to break existing widgets that use the `query` API.
Since it's not an officially documented one I think this is ok.

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (erezrokah) [inline packages/netlify-cms-widget-relation/src/RelationControl.js:243]: What do you think about moving this to https://github.com/netlify/netlify-cms/blob/d057e023640e2076ab8be1f3ce2364a992cdce3e/packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControl.js#L314

File: packages/netlify-cms-core/src/reducers/index.ts
Comment (erezrokah) [inline packages/netlify-cms-core/src/reducers/index.ts]: Seems to be related to https://github.com/immutable-js/immutable-js/issues/1246

**Review by erezrokah — APPROVED** [review]
Thanks @smashercosmo again for another great PR.
I have some comments, but they are mostly relevant to the old code as well (I approved the PR so I'll let you comment before I merge).

After reading the PR I'm convinced that we should separate `query` into its own state.
While both `search` and `query` use similar backend logic, `search` is for the collections view and `query` is only used for the relation widget.

For example, they both share the same `isFetching` and `term` state which I find odd.
Also `entryIds` is only relevant for collections and `queryHits` for the relation widget.

Separating those will allow us to persist user search between screens.

I think it can be done in another PR.

**Comment by smashercosmo** [comment]
> After reading the PR I'm convinced that we should separate `query` into its own state.

Exactly my thoughts while I was working on the PR

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (smashercosmo) [reply packages/netlify-cms-widget-relation/src/RelationControl.js:243]: Not sure I understand the proposal

File: packages/netlify-cms-core/src/reducers/search.ts
Comment (smashercosmo) [reply packages/netlify-cms-core/src/reducers/search.ts:74]: I don't know. Probably. Just to be safe.

File: packages/netlify-cms-core/src/actions/search.ts
Comment (smashercosmo) [reply packages/netlify-cms-core/src/actions/search.ts:113]: Oh, ok, I understand. You are allowed to search for the same thing twice, but you're guarded from pressing 'enter' multiple times if search hasn't finished yet.

File: packages/netlify-cms-widget-relation/src/RelationControl.js
Comment (erezrokah) [reply packages/netlify-cms-widget-relation/src/RelationControl.js]: See https://github.com/netlify/netlify-cms/pull/5157/commits/3a68c43697a683911a9a590e5c760f235f5f805f

Comment (smashercosmo) [reply packages/netlify-cms-widget-relation/src/RelationControl.js]: Hm... according to EditorControl propTypes queryHits can be either array or object

Comment (smashercosmo) [reply packages/netlify-cms-widget-relation/src/RelationControl.js]: So in case of array ```queryHits[this.uniqueFieldId]``` won't work

Comment (erezrokah) [reply packages/netlify-cms-widget-relation/src/RelationControl.js]: Same as `RelationControl` https://github.com/netlify/netlify-cms/blob/3211f94f4ae55cc008bc28325a5387722657cd2c/packages/netlify-cms-widget-relation/src/RelationControl.js#L95

I'll verify it can only be an object

Comment (erezrokah) [reply packages/netlify-cms-widget-relation/src/RelationControl.js]: Yep, I think with https://github.com/netlify/netlify-cms/pull/5157/commits/96e378134ae232d56ed77a13e7f1748dd7e244ca the prop types now match https://github.com/netlify/netlify-cms/pull/5157/files#diff-f141f1b29eb66150ed0215315c55f8e2d91dae5fb1f947c416475c5d2e4a2028R21

**Pull request merged**
